source/common/stats/allocator_impl.cc

```diff
@@ -258,11 +258,46 @@ class TextReadoutImpl : public StatsSharedImpl<TextReadout> {
   std::string value_ ABSL_GUARDED_BY(mutex_);
 };
 
+// An implementation of Stats::CounterGroup.
+class CounterGroupImpl : public StatsSharedImpl<CounterGroup> {
+public:
+  CounterGroupImpl(StatName name, AllocatorImpl& alloc, StatName tag_extracted_name,
+                   const StatNameTagVector& stat_name_tags, size_t max_entries)
+      : StatsSharedImpl(name, alloc, tag_extracted_name, stat_name_tags), max_entries_(max_entries),
+        values_(std::make_unique<std::atomic<uint64_t>[]>(max_entries)),
+        pending_values_(std::make_unique<std::atomic<uint64_t>[]>(max_entries)) {}
+
+  void removeFromSetLockHeld() ABSL_EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) override {
+    const size_t count = alloc_.counter_groups_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void add(size_t index, uint64_t amount) override {
+    // Note that a reader may see a new value but an old pending_increment_ or
+    // used(). From a system perspective this should be eventually consistent.
+    values_[index] += amount;
+    pending_values_[index] += amount;
+    flags_ |= Flags::Used;
```

Comment:
it's worth considering whether we might want to replicate the 'used' bit per entry.

Reply:
I wondered about that... What are the semantics of the used bit? Like, how is it consumed typically?

Reply:
I think it's used for transferring stats across hot-restart generations. Maybe it's fine to have one used-bit for all the stats in the group. But we should socialize this with maintainers.

Reply:
Sounds good. What's the best way to go about socializing this?

Reply:
I'd make a github issue describing the reasons for doing this and we can ask to comment on it in Slack and also bring it up in the community meeting, which happens every 2 weeks. The next meeting is tuesday 5/25 @ 12pm eastern.

Reply:
I guess the other thing this reminds me of: do we want these stats to be transferred across hot-restart generations? I'd guess the answer is 'yes', that's what would happen if they were counters.

So we'll need to add explicit code to do the RPC of the groups across the boundary, and consider corner cases like the set of counters changing between generations.